app/concepts/dossier_entreprise/operation/fetch_immatriculation_principale.rb

```diff
@@ -0,0 +1,50 @@
+class DossierEntreprise
+  module Operation
+    class FetchImmatriculationPrincipale < Trailblazer::Operation
+      step :any_immatriculations_principales?
+      failure :any_immatriculations_secondaires?, Output(:success) => Track(:immatriculation_secondaire)
+      failure :log_not_registered_to_rncs, Output(:failure) => 'End.failure'
+
+      step :uniq?, Output(:success) => 'End.success'
+      failure :none_date_immatriculation_nil?, Output(:success) => Track(:success)
+      failure :log_failed_to_find_valid_immatriculation, Output(:failure) => 'End.failure'
+      step :choose_latest_immatriculation_principale
+
+      failure :log_registered_but_no_immatriculation_principale_found, magnetic_to: [:immatriculation_secondaire]
```

Comment:
Je propose quelque chose d'un peu plus simple : pour que ce soit plus lisible et aussi pour ne faire qu'une unique requête en base de données (important quand on fait tourner des scripts de couverture du périmètre sur les 6 millions de siren). 

WDYT ?

```ruby 
step :any_immatriculation? 
fail :log_no_rncs_immatriculation, fail_fast: true
step :any_immatriculation_principale?
fail :log_immat_secondaire_only, fail_fast: true
step :many_immat_and_some_blank_date_immatriculation?
fail :log_cannot_identify_latest_immat, fail_fast: true
step :return_only_or_latest_immat_principale


def any_immatriculation?
  ctx[:dossiers] = DossierEntreprise.where(siren: siren)
  ctx[:dossiers].any?
end 

def any_immatriculation_principale?
  ctx[:immat_principales] = dossiers.select { |d| d.type_inscription = 'P' }
  ctx[:immat_principales].any?
end

def many_immat_and_some_blank_date_immatriculation?
  many_immat? = immat_principales.count > 1
  blank_date_immat? = immat_principales.none? { |d| d.date_immatriculation.nil? }
  many_immat? && blank_date_immat?
end 
```

Reply:
on ne peut pas mettre de `fail_fast` dans une opération qui est nestée ; ça fait échouer l'opération "parente". C'est pour ça que j'ai utilisé des `failure :log_foo, Output(:success) => 'End.failure'` à la place

Ou sinon il faut que _cette_ opération là renseigne directement le `ctx[http_error]` ce dont je n'ai pas envie car j'aurai aimé que cette opération puisse être "indépendante" du reste

sinon pour la logique j'ai push une version qui fonctionne

Reply:
Si si ça marche `fail_fast` dans une opération nestée 😉 Fondamentalement tu peux tout faire en TRB, mais de manière générale j'essaie d'éviter au maximum les "branchements" dans tous les sens à des fins de lisibilité (d'où ma propal précédente ^^). 

En fait, si ton opération nestée termine en `fail_fast`, l'opération appelante va finir en `fail_fast`, **sauf** si tu fais le branchement que tu veux. Je paste mes exemples en console ci-dessous : 

<img width="617" alt="Capture d’écran 2019-10-30 à 12 21 44" src="https://user-images.githubusercontent.com/4283958/67854316-5ecc3300-fb10-11e9-9cbe-2052aae90bcf.png">

C'est plutôt bien foutue et [documenté](https://2019.trailblazer.to/2.1/docs/activity.html#activity-wiring-api).

Reply:
Du coup je tiens quand même à une version de l'opération proche de celle que j'ai proposé : 
1. c'est quand même plus lisible (là si je veux être sûr du flow parcouru il me faut du papier et un crayon) ; 
2. une seule requête en base (et ça j'y tiens parce que je vais faire une passe d'optimisation, c'est beaucoup trop lent et bourrin en l'état pour que ce soit viable en production).

Pour ce qui est du code HTTP renseigné dans la nested opération je ne suis pas vraiment d'accord. Pour moi si cette opération nestée échoue c'est dans tous les cas un 404, seul le message d'erreur change. Le code peut être set dans l'opération qui appelle.

Reply:
Oublie le message plus haut, j'ai cliqué sur le bouton "View changes" mais visiblement ça n'a pas marché et je ne voyais pas tes changements post review... Au final ton refactor est iso à ce que je proposais, il y a juste des redirections vers 'End.failure' au lieu de 'fail_fast' ce qui est pareil. 

Donc c'est bon pour moi 👍

Reply:
tes remarque m'ont quand même fait penser que je peux utiliser `fail_fast`, faut juste que je rattrape l'opération nestée ainsi :
`step Nested(Foo), Output(:fail_fast) => Track(:failure)` 

Et voilà ! c'est bien plus joli !